### ISSUE TITLE
Cat Along a Shard Axis

### ISSUE THREAD

**Issue by tobias17** [tinygrad/tinygrad#6638]
## The Problem

Originally uncovered in the work of #6328 in support of SDXL inference, this issue deals with the fact that tinygrad fundamentally is not designed for this operation.

The following is a distilled version of a use case one might need this functionality in:
```python
GPUS = [f"{Device.DEFAULT}:{i}" for i in range(4)]
DEVICE_BS = 2
GLOBAL_BS = DEVICE_BS * len(GPUS)

class Model:
  def __init__(self, size:int):
    self.in_1 = Linear(size, size*2)
    self.in_2 = Linear(size, size*2)
    self.out  = Linear(size*2, size)
  def __call__(self, x, cond):
    return self.out(self.in_1(x) + self.in_2(cond))

model = Model(10)
for w in get_state_dict(model).values():
  w.shard_(GPUS, axis=None)

ITERATIONS = 10
x      = Tensor.randn(GLOBAL_BS,10).shard(GPUS, axis=0)
cond_u = Tensor.zeros(GLOBAL_BS,10).shard(GPUS, axis=0)
cond_c = Tensor.randn(GLOBAL_BS,10).shard(GPUS, axis=0)
for i in range(ITERATIONS):
  latent_u, latent_c = model(Tensor.cat(x, x), Tensor.cat(cond_u, cond_c)).chunk(2)
  scaled = latent_u + 8.0 * (latent_c - latent_u)
  x = x + ((1.0 / ITERATIONS) * (x - scaled))
print(x.numpy())
```

The above works fine when running on a single GPU, but when the sharding is introduced it starts causing problems. This is due to `Tensor.cat` not working along the shard axis.

### Cat the Devices

There was some investigation into adding a special case that would cat the devices and lazybuffers. For example 2 tensors with devices `('NV:1','NV:2')` when cat together would produce a MultiLazyBuffer with device `('NV:1','NV:2','NV:1','NV:2')`. While this achieves the purpose of having them together, this makes it no longer possible to perform computations with a weight who's device is `('NV:1','NV:2')` since you get a device mismatch.

### Rewrite for "N Batch Dims" Independence

Another approach that I'm warming up to is going through the framework and SDXL code and rewriting it all to be agnostic to the number of batch dimensions. That is, if a piece of code expects a 3 dim input (something like shape `(4, 128, 128)`), then it should not care how many batch dimensions there are. It should be able to accept shapes like `(4, 128, 128)`, `(15, 4, 128, 128)`, and `(2, 2, 2, 4, 128, 128)`.

The framework has been pretty good about this from what I've seen so far, and is something I focused on when implementing tinygrad's initial `Tensor.interpolate` function. It bothered me that pytorch's version required exactly 2 "batch dimensions", no more no less.

The SDXL code will need some work to get converted over. I had wanted to go through and use `rearrange` in places, but this seems to not support this goal of arbitrary number of batch dimensions. Ideally `rearrange` would get extended to support this paradigm, but this is a far reaching goal.

If the SDXL inference code (and the framework APIs it calls) all get converted to the batch dim independence paradigm, the n solving the problem laid out in this issue becomes fairly trivial. Adding a 2nd batch dimension allows you to call `Tensor.cat` along that new dimension and the inference code would not care if you had 0, 1, 2, 3, or more batch dimensions.
